coderd/rbac/authz.go

```diff
@@ -66,32 +66,37 @@ type RegoAuthorizer struct {
 
 var _ Authorizer = (*RegoAuthorizer)(nil)
 
-// Load the policy from policy.rego in this directory.
-//
-//go:embed policy.rego
-var policy string
-
-const (
-	rolesOkCheck = "role_ok"
-	scopeOkCheck = "scope_ok"
+var (
+	// Load the policy from policy.rego in this directory.
+	//
+	//go:embed policy.rego
+	policy string
+	query  rego.PreparedEvalQuery
 )
 
-func NewAuthorizer() (*RegoAuthorizer, error) {
-	ctx := context.Background()
-	query, err := rego.New(
+func init() {
+	var err error
+	query, err = rego.New(
 		// Bind the results to 2 variables for easy checking later.
 		rego.Query(
 			fmt.Sprintf("%s := data.authz.role_allow "+
 				"%s := data.authz.scope_allow",
 				rolesOkCheck, scopeOkCheck),
 		),
 		rego.Module("policy.rego", policy),
-	).PrepareForEval(ctx)
-
+	).PrepareForEval(context.Background())
 	if err != nil {
-		return nil, xerrors.Errorf("prepare query: %w", err)
+		panic(err)
 	}
-	return &RegoAuthorizer{query: query}, nil
+}
+
+const (
+	rolesOkCheck = "role_ok"
+	scopeOkCheck = "scope_ok"
+)
+
+func NewAuthorizer() *RegoAuthorizer {
+	return &RegoAuthorizer{query: query}
 }
```

Comment:
Can we just do this with a sync.Once instead of a `init()`? I just don't like `inits()` in packages.

```golang
var authorizerOnce = sync.Once{}

// cachedQuery is cached to only be compiled once.
var cachedQuery *rego.PreparedEvalQuery

func NewAuthorizer() (*RegoAuthorizer, error) {
	ctx := context.Background()
	var query rego.PreparedEvalQuery
	var err error
	// Only compile the authorizer once and reuse it.
	authorizerOnce.Do(func() {
		query, err = rego.New(
			// Bind the results to 2 variables for easy checking later.
			rego.Query(
				fmt.Sprintf("%s := data.authz.role_allow "+
					"%s := data.authz.scope_allow",
					rolesOkCheck, scopeOkCheck),
			),
			rego.Module("policy.rego", policy),
		).PrepareForEval(ctx)
		if err != nil {
			err = xerrors.Errorf("prepare query: %w", err)
		}
		cachedQuery = &query
	})
	// Error from the first call
	if err != nil {
		return nil, err
	}
	if cachedQuery == nil {
		return nil, xerrors.Errorf("cached query failed to compile")
	}

	return &RegoAuthorizer{query: *cachedQuery}, nil
}
```

Reply:
I think erroring from the first call will result in more strange behavior... because you'd get an indirect error on the second execution. We panic'd before in `coderd.go`, so I tried to keep the same behavior.

Reply:
Yea, that is annoying. Could cache the error, but panicing is fine